### PR TITLE
Made compilation error colors come through et.

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/147931

environment variable documented in github issue: https://github.com/ninja-build/ninja/issues/2196

## screenshot of results
Screenshot 2024-12-13 at 10 18 15 AM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/engine/blob/main/docs/testing/Testing-the-engine.md
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md

### CODE REVIEW

**Comment by gaaclarke** [comment]
Ahhhh, nertz!  This breaks the gcc path fix since it breaks the regular expression.  I'm putting this back into Draft and removing reviewers.

File: tools/pkg/engine_build_configs/lib/src/build_config_runner.dart
Comment (matanlurey) [inline tools/pkg/engine_build_configs/lib/src/build_config_runner.dart:641]: I wonder if this is correct, that is, if this should be conditionally set _if_ the current environment is an ANSI terminal with color support. 

If it was conditionally set, would we need the parsing and massaging you did above?

**Review by matanlurey — CHANGES_REQUESTED** [review]

**Review by zanderso — COMMENTED** [review]
It's good practice with colors to have some way to disable them on the command line. Is there a `ninja` command line option (that could be plumbed into `extraNinjaArgs` in another PR) that will override the environment variable?

File: tools/pkg/engine_build_configs/lib/src/build_config_runner.dart
Comment (gaaclarke) [reply tools/pkg/engine_build_configs/lib/src/build_config_runner.dart:641]: The behavior most people follow is if you aren't connected to a tty you don't emit ansi escape sequences.  That's the behavior ninja was doing.  I can emulate it.  It's only a problem if someone wants to run et as a subprocess.

**Comment by gaaclarke** [comment]
> It's good practice with colors to have some way to disable them on the command line. Is there a `ninja` command line option (that could be plumbed into `extraNinjaArgs` in another PR) that will override the environment variable?

I'll duplicate ninja's behavior and use the environment variable.  One sec, it's also related to matans comment.

File: tools/pkg/engine_build_configs/lib/src/build_config_runner.dart
Comment (gaaclarke) [reply tools/pkg/engine_build_configs/lib/src/build_config_runner.dart]: Done, now we only show the colors if stdout supports it or if it has been overwritten with `CLICOLOR_FORCE`.

**Comment by gaaclarke** [comment]
> It's good practice with colors to have some way to disable them on the command line. Is there a `ninja` command line option (that could be plumbed into `extraNinjaArgs` in another PR) that will override the environment variable?

Done, see https://github.com/flutter/engine/pull/57174#discussion_r1884341434

File: tools/pkg/engine_build_configs/lib/src/build_config_runner.dart
Comment (gaaclarke) [reply tools/pkg/engine_build_configs/lib/src/build_config_runner.dart]: > If it was conditionally set, would we need the parsing and massaging you did above?

Yes, we still need it because the gcc path fix is a whole different thing that assumed no ansi escape sequences.

**Review by matanlurey — APPROVED** [review]
Seems reasonable to me. Please get an explicit approval from zra@ before merging.

File: tools/pkg/engine_build_configs/lib/src/build_config_runner.dart
Comment (zanderso) [inline tools/pkg/engine_build_configs/lib/src/build_config_runner.dart:637]: There should be a way to disable color even when attached to an ansi terminal. It doesn't have to be plumbed all the way to the `et` command line in this PR, but it should at least be exposed, for example as an argument to the constructor of the containing class.

Comment (gaaclarke) [reply tools/pkg/engine_build_configs/lib/src/build_config_runner.dart]: Done, I made the negation turn off colors too despite them being supported.  I'm not sure of the use-case where they are supported but someone wouldn't' want them.  Usually they are disabled to support parsing.

**Review by zanderso — APPROVED** [review]

**Pull request merged**
